### ISSUE TITLE
Wrong extension and MIME type with conditionally served images in the 'Save Image As...' dialog

### ISSUE THREAD

**Issue by rhinoduck** [MoonchildProductions/Pale-Moon#959]
I ran into this while adding JXR support, but as it not limited to JXR, I am opening this issue. 

On Windows, the 'Save Image As...' dialog offers to save images with the wrong extension and MIME type if their conditionally served counterpart MIME type is present in Windows registry under `HKEY_CLASSES_ROOT\MIME\Database\Content Type`.

`image/png` and `image/vnd.ms-photo` are the two MIME types that are present by default and that could be problematic in Pale Moon. `image/png` because the current Accept header in Pale Moon looks like this: `image/webp,image/png,image/*;q=0.8,*/*;q=0.5`; and `image/vnd.ms-photo` because of how servers might be configured due to MS Edge and IE advertising `image/jxr` but only accepting `image/vnd.ms-photo`.

I don't know if, by installing a program or changing Windows configuration, more types could appear, but if an entry for WebP appeared, that would be a problem too.

Demonstration:
(follow the link and trigger the 'Save Image As...' action from the context menu)

1) [image/png](http://rhinoduck.net/palemoon/github/issues/105/conditional-serving/png/image.html)  - observable in current stable version of Pale Moon (27.1.2)
2) [image/vnd.ms-photo](http://rhinoduck.net/palemoon/github/issues/105/conditional-serving/edge/image.html) - observable in current Pale Moon dev build with JXR support on

In the above cases, jpg should be offered, but png/wdp is wrongfully offered; the actual data saved will be JPEG data because it will come from a separate HTTP request which is not an image request (and thus does not include the additional image MIME types in the Accept header that the image requests include).

Some screenshots to quickly illustrate the issue and the difference between Pale Moon and Firefox:

Pale Moon 27.1.2 in the PNG case
![pm](https://cloud.githubusercontent.com/assets/7115791/23980666/6e2e0f88-0a01-11e7-8ebe-59e752b8ddcd.png)

Firefox 51.0.1 in the PNG case (with `image.http.accept` amended to contain `image/png`)
![ff](https://cloud.githubusercontent.com/assets/7115791/23981113/7aafbca4-0a04-11e7-9ef4-5757cc329482.png)

I remember from when looking into this that the problem came down to the extension and MIME type being pulled from cache and nsExternalHelperAppService and the surrounding JavaScript having to decide based on conflicting MIME type and extension (jpg extension, but png/ms-photo MIME type) and preferring the primary extension for the MIME type found in the Windows registry (if found) which will be wrong in these cases; also note that no request to the server is made during this time, so the information from the cache is the only information available.

Maybe this can simply be resolved by changing how nsExternalHelperAppService picks what gets offered in the end, but I didn't feel like I had enough knowledge about why things are the way they are to avoid possibly breaking something. So I am hoping someone with better understanding of the logic behind these parts of code and knowledge of the history surrounding them can pick it up. If that doesn't happen, I may come back to it sometime in the future.

---

This has stopped happening in my main profile after I upgraded from 27.1.1 to 27.1.2 (which makes no sense), but is still happening in other profiles (including freshly created profiles). Can someone try the PNG case and confirm that there is actually somebody else out there seeing what I am seeing (i.e. that which is in the first screenshot)?

**Comment by JustOff** [comment]
> Can someone try the PNG case and confirm that there is actually somebody else out there seeing what I am seeing (i.e. that which is in the first screenshot)?

I confirm.

**Comment by wolfbeast** [comment]
We can probably see how Mozilla worked around this conflict. If someone could run mozregression and pinpoint the BMO bug, we can see what they did there without inventing the wheel twice.

**Comment by wolfbeast** [comment]
I think what should happen is using "save image as" that the "image" http-accept header should be sent -- this is probably not the case now and a raw http request (with the generic accept header) is used instead, even if it is a request for an image.

**Comment by wolfbeast** [comment]
Looks like this bug has been in Firefox code since forever (back to 0.8)
https://bugzilla.mozilla.org/show_bug.cgi?id=241639

**Comment by JustOff** [comment]
Here is pushlog I found using mozregression: https://hg.mozilla.org/mozilla-central/pushloghtml?fromchange=8ef94be995a453f5c464278c53478ba8c8554f81&tochange=2b5237c178ea02133a777396c24dd2b713f2b8ee

**Comment by JustOff** [comment]
Probably https://bugzilla.mozilla.org/show_bug.cgi?id=1249474

**Comment by wolfbeast** [comment]
Switching it to */* will simply break server-side conditional sending of image formats (and we'll lose our advantage of having servers actually serving supported formats in lieu of generic fallback ones). It doesn't solve this issue which is a different issue altogether and only pops up because a mime type will NOT be specified in that case.
We need to specify `image/webp` and `image/jxr` to let servers know we support these formats, since otherwise they will assume only generally-accepted/supported formats like jpg, png, gif.

**Comment by wolfbeast** [comment]
Adding image/png to Firefox's `image.http.accept` will result in the exact same behavior as seen on Pale Moon. There is 0 difference with how this is handled.

**Comment by wolfbeast** [comment]
Poked Mozilla about this in https://bugzilla.mozilla.org/show_bug.cgi?id=1347958

**Comment by rhinoduck** [comment]
> Adding image/png to Firefox's image.http.accept (as stated in the initial report) will result in the exact same behavior as seen on Pale Moon. There is 0 difference with how this is handled.

I did actually also check the Firefox code back when I was looking at this in the Pale Moon code and came to the conclusion that FF would have the same issue, so I was surprised when I saw what you can see in the second screenshot. It still looks like that for me in Firefox and in my main Pale Moon profile since I upgraded to 27.1.2; I have no idea what causes that.

---

Btw, this might be another relevant bug to bring it up in (and one where people might actually be interested in it) since the question of advertising WebP support in the Accept header of image requests already came up there: https://bugzilla.mozilla.org/show_bug.cgi?id=1294490

**Comment by rhinoduck** [comment]
I am pasting here the Apache 2 configuration I use for the `png` directory on the linked test site so that all information is available and so that it can be checked for mistakes.

```
<FilesMatch "\.(jpe?g)$">
    RewriteCond %{HTTP_ACCEPT} image/png
    RewriteRule (.+)\.[^.]+$ $1.png [T=image/png]
    Header append Vary Accept
</FilesMatch>
<FilesMatch "\.png$">
    Header append Vary Accept
</FilesMatch>
```

**Comment by wolfbeast** [comment]
What you show in your screenshot is what would happen if image.http.accept is Firefox's default of `*/*`, not with `image/png` added to it. Did you add it in front, with a comma?

I don't think that bringing it up in the WebP implementation bug at mozilla is going to be useful. That is about adding the image support; the http accept header has already been discussed there, and this issue is not otherwise related. At most a "see also".

**Comment by rhinoduck** [comment]
> What you show in your screenshot is what would happen if image.http.accept is Firefox's default of */*, not with image/png added to it. Did you add it in front, with a comma?

`image/png,*/*` is what I modified `image.http.accept` in Firefox to. Notice the `*.png` at the end of `Save as type` in the second screenshot, that would not have been there have I not done so. Besides that, I wouldn't have got back the PNG image in the document in the first place, so that makes me confident enough I got that right.

As I mentioned before, I also get this in my main Pale Moon profile where I did not touch the pref at all. This is some other kind of strangeness. I'm not sure we are going to get anywhere unless someone else has this happen or unless I am able to reproduce it on command. Anyway, this is a complication I'd disregard as long as everyone else is consistently getting the results from screenshot 1. In the previous post, I was mostly just communicating my surprise.

> I don't think that bringing it up in the WebP implementation bug at mozilla is going to be useful.

The reasoning behind this was that it will directly affect WebP if they decide to advertise it. Besides this, I don't think there will be any incentive for anyone to look into it because as long as there is only `*/*` in the Accept header for image requests, the issue simply doesn't exist.

But it was just a thought, I bet you know the ways of BMO to an immensely greater extent than I.
